packages/webpack/lib/setEnv.js

```diff
@@ -0,0 +1,37 @@
+module.exports = setEnv
+
+/**
+ * Replaces constants in your source code with values from `process.env`
+ * using the `webpack.EnvironmentPlugin`.
+ *
+ * Using `setEnv` multiple times results in a single
+ * EnvironmentPlugin instance configured to do all the replacements.
+ *
+ * @param {string[]|object} constants
+ * @return {Function}
+ */
+function setEnv (constants) {
+  const setter = context => prevConfig => {
+    context.setEnv = Object.assign({}, context.setEnv, toObject(constants))
+    return prevConfig
+  }
+
+  return Object.assign(setter, { post: addEnvironmentPlugin })
+}
+
+function addEnvironmentPlugin (context, util) {
+  return util.addPlugin(
+    new context.webpack.EnvironmentPlugin(context.setEnv)
+  )
+}
+
+function toObject (constants) {
+  if (Array.isArray(constants)) {
+    return constants.reduce((result, constant) => {
+      result[constant] = undefined
```

Comment:
I think this place needs an explanation. It’s not clear what we need this unless you know how EnvironmentPlugin works.

Reply:
Don't know how you feel about it, but maybe we shouldn't rely on that implicit magic and rather go for an explicit default here in the `setEnv()` block. Also it looks more like a hack than a documented feature, so I would fear that this behavior might break with a future update.

Reply:
It is documented an we’ve discussed that already ;-)

But it supposed to be `null` as I understand.

Reply:
There is nothing magical in here, EnvironmentPlugin resolves arrays in a very similar fashion. The reason it's `undefined` is to have the following warning:
```
WARNING in EnvironmentPlugin - BABEL_ENV environment variable is undefined.

You can pass an object with default values to suppress this warning.
See https://webpack.js.org/plugins/environment-plugin for example.
```
Defaulting to `null` will swallow the warning which is not useful.

Reply:
Ahhh, I think I just got the idea. We normalize arrays to objects to make it all composable. Setting the default value to `undefined` achieves that, setting to `null` does not, since the `EnvironmentPlugin` would think that `null` is the intended default value.
Right?